operators/pkg/controller/elasticsearch/driver/downscale_invariants.go

```diff
@@ -0,0 +1,70 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package driver
+
+import (
+	"github.com/elastic/cloud-on-k8s/operators/pkg/apis/elasticsearch/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/reconcile"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/operators/pkg/utils/k8s"
+	appsv1 "k8s.io/api/apps/v1"
+)
+
+const (
+	OneMasterAtATimeInvariant        = "A master node is already in the process of being removed"
+	AtLeastOneRunningMasterInvariant = "Cannot remove the last running master node"
+)
+
+// DownscaleInvariants restricts downscales to perform in a single reconciliation attempt:
+// - remove a single master at once
+// - don't remove the last living master node
+type DownscaleInvariants struct {
```

Comment:
To add to that, name suggests that it's applicable to downscales and while true, it's imprecise as it applies only to downscales of master ssets.

Reply:
Hmm at first I thought that was okay, because we might extend this later to include things other than masters. But I'm not sure what those might be.

Reply:
Yes: one good example I was hesitant with: don't remove the last data node of the cluster. May happen if, for example, you mutate one NodeSpec to another with a different name. We spawn a new statefulset and downscale the current one.
We used to have this in the code but since now we rely on StatefulSets it may not be as important. I think I'd still like to keep the struct name generic to also apply to any non-master node.

Reply:
> Naming nit: this struct is not representing the invariants but accumulates the state of their violation ...

+1. `invariants` suggests some kind of stateless thing, whereas the current struct is clearly stateful. 
I could separate state (`find-a-name`) and logic (`invariants`) for a cleaner separation of concerns (functional vs. object-oriented), but it probably does not change the code much in the end (still have to propagate state into the functions).

Do you have an other naming suggestion? 
Maybe have an `ongoingDownscale` struct for the state (with a `recordOneRemoval ` function), that we pass to stateless funcs?

Reply:
Not really. As I said it's a nit so it should not delay this PR. The only thing I can think of is to add a suffix like `State` to the struct to make it clear what is going on here.

Reply:
https://github.com/elastic/cloud-on-k8s/pull/1549/commits/8b59ee220f00695abe499aa7e2e9a329168c08cc should fix this :)